### ISSUE TITLE
twisted.python.modules.walkModules fails completely if there is an unimportable package anywhere in the path

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3419]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3419|
|Type|defect|
|Created|2008-09-07 14:36:04Z|
|Branch|https://github.com/twisted/twisted/tree/walkModulesNone-3419|

```
exarkun@boson:/tmp/modules$ python
Python 2.5.2 (r252:60911, Jul 31 2008, 17:28:52) 
[GCC 4.2.3 (Ubuntu 4.2.3-2ubuntu7)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from sys import modules
>>> modules['OpenSSL'] = None
>>> from twisted.python.modules import walkModules
>>> list(walkModules())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/modules.py", line 715, in walkModules
    for module in package.walkModules(importPackages=False):
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/modules.py", line 175, in walkModules
    for package in self.iterModules():
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/modules.py", line 127, in iterModules
    for placeToLook in self._packagePaths():
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/modules.py", line 420, in _packagePaths
    load = self.load()
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/modules.py", line 380, in load
    return self.pathEntry.pythonPath.moduleLoader(self.name)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/reflect.py", line 460, in namedAny
    raise ModuleNotFound("No module named %r" % (name,))
twisted.python.reflect.ModuleNotFound: No module named 'OpenSSL'
>>> 
```

Specifically, this makes the unit test for `walkModules` rather fragile.  In trying to resolve [#3111](https://github.com/twisted/twisted/issues/3111), it is not clear how to make this test pass.

Generally, it's kind of crummy for any one bad module in the path to completely disable this API.

Also, it's kind of crummy that the `twisted.python.modules` tests load such a huge quantity of Python code, since this makes them rather slow.

<details><summary>Searchable metadata</summary>

```
trac-id__3419 3419
type__defect defect
reporter__exarkun exarkun
priority__high high
milestone__ 
branch__branches_walkModulesNone_3419 branches/walkModulesNone-3419
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1220798164000000 1220798164000000
changetime__1274655983000000 1274655983000000
version__None None
owner__ 
cc__jesstess cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

Updating the summary because it's actually only un-importable _packages_ which cause problems.

Also, I tried to make a package with a syntactically invalid `__init__.py` and that didn't trigger the problem; it's something to do with the case which involves the presence of a `None` in `sys.modules`.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph removed owner<br>@glyph set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

(In [28722]) Branching to 'walkModulesNone-3419'

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_7"></a>|@jesstess set owner to @glyph|
|-|-|

test_modules.py
 * needs a copyright bump
 * There are some assertEqual in test_nonexistentPaths and test_nonDirectoryPaths that should be assertEquals
 * In the docstrings for the two new tests: "If a module has been blacklisted by setting a C{None} key in sys.modules" sounds like it means the same things as "If a package has been explicitly forbidden from importing by setting it to C{None} in sys.modules". If they do mean the same thing, sticking to one description would be clearer.
 * test_unimportablePackage says "attempting to walk over it should not load it". Does the existing assertion take care of checking if test_package was loaded, or does there need to be another assert?
 * os is imported but unused

modules.py
 * needs a copyright bump
 * in PythonPath.__getitem__, the modname parameter needs epytext markup
 * in PythonPath.__getitem__, the @raises statement looks like it's missing a negation.

[builds](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/walkModulesNone-3419) look good.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

(In [28991]) Review feedback for test_modules.

    * copyright bump
    * assertEqual -> assertEquals
    * Fixed docstrings, logic, and names for the two unimportable-package tests to be consistent with each other, and make it clear that they're testing similar functionality of two different APIs.
    * added 'is loaded' assertion to test_unimportablePackageWalkModules
    * removed unused os import

Refs: [#3419](https://github.com/twisted/twisted/issues/3419)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

(In [28992]) Review feedback for twisted.python.modules.

    * copyright bump
    * `PythonPath.__getitem__`: a better docstring, with
        * `@param` & `@type` for `modname`
        * `@raise` with a more helpful message (and corrected epytext syntax)
        * separated `@return` and `@rtype`

Refs: [#3419](https://github.com/twisted/twisted/issues/3419)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph removed owner|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/walkModulesNone-3419)

I think that should be it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set owner to @glyph|
|-|-|

Needs a news fragment.  Otherwise looks good, please add that and merge.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph set status to `closed`|
|-|-|

(In [29007]) Merge walkModulesNone-3419

Author: glyph

Reviewer: exarkun, jesstess

Fixes: [#3419](https://github.com/twisted/twisted/issues/3419)

twisted.python.modules.walkModules will now handle packages explicitly precluded from importing by a None placed in sys.modules.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
